### ISSUE TITLE
Build Failing Trying to Pull Docker Image

### ISSUE THREAD

**Issue by CUCraigT** [turnkeylinux/tracker#1502]
I am attempting to build an appliance to basically just run Docker and a couple Docker containers. I have successfully gotten Docker installed based on TKL Core and installing those packages. However, I then want to pull the two Docker images I want using `docker pull`. This fails during the Make looking like the socket is not available. I have even tried to start the the docker service explicitly and still no luck. Here is the error I get:

```
+ echo Start Docker Service
Start Docker Service
+ systemctl start docker
[ ok ] Starting Docker: docker.
+ whoami
root
+ echo Pull Docker Images
Pull Docker Images
+ docker pull mcr.microsoft.com/mssql/server:2017-latest
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
make: *** [/usr/share/fab/product.mk:540: build/stamps/root.patched] Error 1
```

Here is my `conf.d` file:

```
#!/bin/sh -ex
SQL_VERSION="2017-latest"
echo "Install Docker"

# add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable"

echo "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable" >> /etc/apt/sources.list.d/sources.list

curl -fsSL https://download.docker.com/linux/debian/gpg | apt-key add -

apt-get update
apt-get install docker-ce docker-ce-cli containerd.io

echo "Start Docker Service"
systemctl start docker

whoami
echo "Pull Docker Images"

#Pull MSSQL Server Image
docker pull mcr.microsoft.com/mssql/server:$SQL_VERSION
```

Any thoughts or tips on how I could do this?

**Comment by JedMeister** [comment]
As per the error message, is the docker daemon running?
```
systemctl status docker
```

Also FWIW [here](https://github.com/turnkeylinux/buildtasks/blob/master/bin/docker-setup) is the install script we use to install docker. It includes additional "best practice" security measures such as not including the repo key as a default "trusted key" and pinning the repository. It's possibly somewhat irrelevant now as you've added the key to your "trusted keys" but the pinning is still a good idea.

**Comment by CUCraigT** [comment]
Thanks for the quick response. I will give your script a try. I did add `systemctl status docker` to the conf.d/main file and got an error saying `status` was unknown:

```
+ echo Start Docker Service
Start Docker Service
+ systemctl start docker
[ ok ] Starting Docker: docker.
+ whoami
root
+ echo Docker Service Status
Docker Service Status
+ systemctl status docker
FATAL [/usr/local/bin/systemctl]: Unknown command 'status'
make: *** [/usr/share/fab/product.mk:540: build/stamps/root.patched] Error 1
```

**Comment by CUCraigT** [comment]
Decided to do `service docker status` instead and it returned that Docker is running. Tried a `docker version` and get the error on the server piece. So, yes it appears the service is running but issues connecting to it. I have seen this when I try to run the commands as non-root, but the build seems to be running as root:

```
+ echo Start Docker Service
Start Docker Service
+ systemctl start docker
[ ok ] Starting Docker: docker.
+ whoami
root
+ echo Docker Service Status
Docker Service Status
+ service docker status
[ ok ] Docker is running.
+ docker version
Client: Docker Engine - Community
 Version:           19.03.12
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        48a66213fe
 Built:             Mon Jun 22 15:45:50 2020
 OS/Arch:           linux/amd64
 Experimental:      false
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
make: *** [/usr/share/fab/product.mk:540: build/stamps/root.patched] Error 1
```

**Comment by JedMeister** [comment]
Ah ok. Apologies, in my haste, I missed that you were building an appliance... (For some reason I just thought that you were trying to install Docker on Core; rereading it now seems obvious that you're creating an appliance...).

So with that context in mind, I suspect that it's a limitation of our build system. Although I'm not 100% sure which one...

We use OverlayFS in the build process (the "decks" that are the `make` targets are overlaid over each other as "layers"). As Docker also uses OverlayFS (AFAIK?), that may be a factor?

The second, and probably more likely, is that our build environment is a chroot. Previous init systems would happily run within a chroot, but that is not the case for systemd. Other than 'enable' and 'disable' (i.e. to enable and disable a service autostart at boot) commands will fail. Some old services provided an SysVinit script (i.e. a start/stop script within `/etc/init.d`) and those can be leveraged via a direct call, e.g. `/etc/init.d/script_name start` or often a `service` call, e.g. `service script_name start`. (note that `service` calls will be picked by `systemctl` when the system is running "properly", but during build time will leverage init.d script when available).

However, what makes all this a somewhat moot point, is that to work around various issues with specific appliances, we actually provide build time wrappers for both the `service` and `systemctl` commands. You can find them in the [`common` - `systemd-chroot` overlay](https://github.com/turnkeylinux/common/tree/master/overlays/turnkey.d/systemd-chroot/usr/local/bin). Note that there is some redundancy between the 2 scripts (I created one and my colleague Stefan created the other).

So what I would suggest is rather than trying to work out what is going on via the script, instead try resolving the issue interactively within the chroot. I.e. from within your appliance build code dir; `/turnkey/fab/products/YOUR_APP`:

```
fab-chroot build/root.patched
```
You will now be inside the build system. You can run commands interactively and test stuff out. Once you're done, you can `exit` to get back to the TKLDev host.

My suspicion is that you'll either want to start the required process manually (in your conf script - don't forget to capture the PID and kill it when you're done). Or alternatively, tweak the `service` wrapper script to do that (and explicitly copy it out and add it to the overlay). You'll find the `service` and `systemctl` wrapper scripts within the chroot at `/usr/local/bin/`.

As hinted above, don't forget to copy out any files that you want to keep. E.g. if you tweak the `service` script. Otherwise any modifications you made will be lost as soon as you re-run `make`.

**Comment by CUCraigT** [comment]
Thank you so much for the detailed response, and your assumption is correct. I am trying to build a simple appliance with Core+Docker+My Docker Image, so we can ship our Docker container to folks that don't have a Kubernetes or Docker environment.

Appreciate the command to get into chroot interactively; I had been struggling to figure that out from the docs. Seeing what is happening is extremely helpful.

With that being said, looks like you **ARE** correct in your assumption with the OverlayFS. When I just try to start the Docker Daemon with /usr/bin/dockerd, I in fact get error trying to mount the overlays:

```
ERRO[2020-08-05T17:52:15.756556350Z] failed to mount overlay: invalid argument     storage-driver=overlay2
ERRO[2020-08-05T17:52:15.762181148Z] AUFS was not found in /proc/filesystems       storage-driver=aufs
ERRO[2020-08-05T17:52:15.763081799Z] failed to mount overlay: invalid argument     storage-driver=overlay
ERRO[2020-08-05T17:52:15.802698034Z] Udev sync is not supported. This will lead to data loss and unexpected behavior. Install a more recent version of libdevmapper or select a different storage driver. For more information, see https://docs.docker.com/engine/reference/commandline/dockerd/#storage-driver-options  storage-driver=devicemapper
```

Going to play around with it a bit more, but have you ever done something similar getting Overlay to work on top of Overlay? The good thing is that Docker works great on my appliance; just really want to be able to seed it with the container images prior to shipping, so it works. If not, I may have to get a startup script on the actual appliance to do the load

**Comment by JedMeister** [comment]
Glad we're on the same page and that my post was useful. :smile: 

I wonder if it's worth trying to see if overlayFS and AUFs will play nice together? We use OverlayFS as it's included in the kernel. However Debian have an AUFs kernel DKMS module package. I can't guarantee it'll work, but it might? Also Docker has a vfs driver. I don't know much about it and apparently it's performance is poor, but for the purposes of building it might be sufficient? you'd probably only want to use the alternate driver at build time though.

**Comment by CUCraigT** [comment]
Worth trying @JedMeister. I'll take a look. I may also take a look at including the docker images as tar.gz files in an overlay directory and then doing a `docker load` command as part of `inithooks`. Will play with that a bit today as well. Thanks for the help working through this. This appliance should be pretty straightforward; just want to make sure the essential files are prepopulated, so they don't **HAVE** to update to get it running.

**Comment by JedMeister** [comment]
> I may also take a look at including the docker images as tar.gz files in an overlay directory and then doing a docker load command as part of inithooks.

That may be a workable optiion. Although FWIW, for our appliances, we aim to have all install and initial config done at build time (then re-initialized as relevant on firstboot; e.g. secret regeneration, etc). Our experience suggests that this approach tends to lead to the fewest possible points of breakage for end users. Although obviously it doesn't eliminate chance of issues, it does reduce vectors.

> just want to make sure the essential files are prepopulated, so they don't **HAVE** to update to get it running.

Good plan! :smile:

**Comment by CUCraigT** [comment]
@JedMeister Got to looking at this again today. Apparently, if you specify one overlay filesystem to pull images and/or build containers and then change it, then those images become inaccessible, so couldn't specify one filesystem at build time and then switch it or we would lose those images (per this doc: https://docs.docker.com/storage/storagedriver/select-storage-driver/)

```
When you change the storage driver, any existing images and containers become inaccessible. This is because their layers cannot be used by the new storage driver. If you revert your changes, you can access the old images and containers again, but any that you pulled or created using the new driver are then inaccessible.
```

Any other options you can think of or am I stuck with `inithooks`, even understanding it is not best practice?

**Comment by JedMeister** [comment]
Argh!

Given that info, my suggestions is to give [aufs](https://docs.docker.com/storage/storagedriver/aufs-driver/) a try?! TBH, I can't be sure how auFS will interact with OverlayFS, and whether it'll work. But I guess there's only one way to find out...

Be aware, that you'll need to install that on both the host (for usage during build time) and the guest (for usage when installed). IIRC these are the packages that you'll need (install on TKLDev for use at build time and add to plan for later usage):

```
aufs-dkms dkms linux-headers-amd64
```
I suspect that you may also want `aufs-tools` installed on the guest too (but no harm installing on the host as well); but I'm not 100% sure.

Sorry I don't have a more precise answer...

**Comment by CUCraigT** [comment]
Thanks, @JedMeister. Just to be clear would this be using `aufs` for TKL or for Docker? I am assuming for Docker but just wanted to clarify.

**Comment by JedMeister** [comment]
@CUCraigT 

> Just to be clear would this be using `aufs` for TKL or for Docker?

TBH, you could go either way. TKLDev historically used to use auFS (it was the default union filesystem provided by Debian). IIRC that was v14.x and earlier. Once overlayFS was merged into the kernel mainline, Debian dropped default auFS support (Debian were patching their kernel to provide it). IIRC that was Debian Stretch (v15.x). When we first started development for that release, we initially just installed auFS in TKLDev as I noted above. However, we ended up refactoring to leverage overlayFS (thus reducing the size a bit). I'm pretty sure the main thing that changed was [`deck`](https://github.com/turnkeylinux/deck). Previously it was a [python script](https://github.com/turnkeylinux/deck/tree/v1.0) but Alon rewrote it in bash for overlayFS support. It would probably be possible to rebuild the old python (2) package and it should work on v16.x, but it's not something I have time for ATM. There may be other bits and pieces.

**TL;DR** I anticipate that using Docker with auFS would be much less work (if it works at all that is). But would conversely increase the size of your resulting ISO and would mean that Docker in the resulting system would need to continue to use auFS.

One other thought that I've had is that IIRC one of the shortcomings of overlayFS is that it only allows a singe read/write layer. I suspect that is the underlaying cause of the issue you're experiencing. Assuming that's the case, then it may be possible to tweak [`fab`](https://github.com/turnkeylinux/fab) to remount `root.patched` as read-only while the docker stuff happens, then remount as read/write once the docker stuff is done?!

**Comment by CUCraigT** [comment]
Interesting. A bit above my head, but I may try to figure that out. Really appreciate the help/ideas.

So, all of Docker stores the images in `/var/lib/docker`. So I just did a `docker pull` on my TKLDEV box and then copied `/var/lib/docker` all to the overlay directory. It actually worked and built an ISO! However, I did some cleanup and tried to build again, and now I am getting the following errors on some of those files when I do a `make` (same error with `make clean`).

```
cp: cannot create special file 'build/root.patched/var/lib/docker/overlay2/a01defc16e24fbcc849141a3492b71d00ba7b21f410c259c9194a0dab8797a4a/diff/etc/nginx/conf.d/default.conf': Operation not permitted
```

There are other files also; just including one so you can see. Any idea, @JedMeister?

**Comment by JedMeister** [comment]
> So, all of Docker stores the images in `/var/lib/docker`. So I just did a docker pull on my TKLDEV box and then copied `/var/lib/docker` all to the overlay directory. It actually worked and built an ISO!

Ah, cool hack! I like the way you think! :smile: 

> However, I did some cleanup and tried to build again, and now I am getting the following errors on some of those files when I do a `make` (same error with `make clean`).

Hmm, that's weird that it worked once, but now won't allow you to rerun any of the steps...

Also the message suggests that it's the copy (`cp`) command which is failing - but you said this occurs when you run `make` and/or `make clean` right? Or did I miss something?!

Although it's not ideal, when all else fails on TKLDev, I suggest a reboot. Then log back in, cd to your build code dir and re-run `make clean`.

As to how to "properly" copy in the `/var/lib/docker` dir from the host, I suggest not copying at all, and instead bind mounting it into the overlay directory. I.e.

```
cd YOUR_BUILD_CODE_DIR
mkdir -p overlay/var/lib/docker
mount --bind /var/lib/docker overlay/var/lib/docker
```

TBH, I'm not 100% sure that will work, but it should...

**Comment by CUCraigT** [comment]
Working on building with `aufs` and ran across another interesting thing @JedMeister. Have looked but haven't found anything.

So, I just do a build and then trying to play around interactively. I run this command, as you stated above:

```
fab-chroot build/root.patched
```
Used to work. Now, even when doing a build on core even in a new TKLDEV VM, I get the following errors.

```
root@tkldev .../products/core# fab-chroot build/root.patched
mount: /turnkey/fab/products/core/build/root.patched/proc: mount point does not exist.
Traceback (most recent call last):
  File "/usr/lib/fab/wrapper.py", line 34, in <module>
    CliWrapper.main()
  File "/usr/lib/python2.7/dist-packages/pyproject.py", line 388, in main
    commands.run(name, args)
  File "/usr/lib/python2.7/dist-packages/pyproject.py", line 261, in run
    command.module.main()
  File "/usr/lib/fab/pylib/cmd_chroot.py", line 110, in main
    chroot = Chroot(newroot, environ=get_environ(env_conf))
  File "/usr/lib/python2.7/dist-packages/chroot.py", line 69, in __init__
    self.magicmounts = MagicMounts(self.path)
  File "/usr/lib/python2.7/dist-packages/chroot.py", line 28, in __init__
    self.mount()
  File "/usr/lib/python2.7/dist-packages/chroot.py", line 39, in mount
    executil.system("mount -t proc", "proc-chroot", self.paths.proc)
  File "/usr/lib/python2.7/dist-packages/executil.py", line 56, in system
    raise ExecError(command, exitcode)
executil.ExecError: non-zero exitcode (32) for command: mount -t proc 'proc-chroot'  '/turnkey/fab/products/core/build/root.patched/proc'
```

ISO builds, but can get in interactively. Am I doing something wrong or missing something all of a sudden?

**Comment by CUCraigT** [comment]
Thanks, @JedMeister. Will try that route a bit more.

Also, tried the `aufs` route tonight. Seems that docker and the filesystem are not seeing it with I go into `fab-chroot build/root.patched`. Get the following error starting dockerd:

```
ERRO[2020-08-21T04:03:05.845580521Z] AUFS was not found in /proc/filesystems       storage-driver=aufs
```

Also, if I cat `/proc/filesystems`:

```
root@tkldev /# cat /proc/filesystems 
nodev	sysfs
nodev	rootfs
nodev	ramfs
nodev	bdev
nodev	proc
nodev	cpuset
nodev	cgroup
nodev	cgroup2
nodev	tmpfs
nodev	devtmpfs
nodev	debugfs
nodev	tracefs
nodev	securityfs
nodev	sockfs
nodev	dax
nodev	bpf
nodev	pipefs
nodev	hugetlbfs
nodev	devpts
nodev	mqueue
nodev	pstore
	ext3
	ext2
	ext4
nodev	autofs
nodev	overlay
nodev	binfmt_misc
root@tkldev /# 
```

I did install `aufs-dkms dkms linux-headers-amd64 aufs-tools` in host and guest.

**Comment by JedMeister** [comment]
The `fab-chroot build/root.patched` will only work if that's the target you have built to (either by design, i.e. `make root.patched`; or by circumstance, i.e. the build failed while building `root.patched`).

If the build completed successfully (either all the way, or to a later build target), then the `build.root.patched` directory will just be an empty directory. Hence the `mount: /turnkey/fab/products/core/build/root.patched/proc: mount point does not exist.` message - because there is no `/turnkey/fab/products/core/build/root.patched/proc` directory.

So if you want to enter the build environment interactively, then you have 2 options.

1. Just build to `root.patched`:
```
make root.patched
```
Or force the build to fail:
```
echo "exit 1" >> conf.d/main
```
(And remove the `exit 1` line from the `conf.d/main` script later on).

2. Build to completion and enter the "sandbox" (`build/root.sandbox`) instead:
```
fab-chroot build/root.sandbox
```
FWIW the sandbox was designed to be a way to quickly tweak an iso. But IIRC it got broken by the change from auFS to OverlayFS. It still exists, but doesn't really work how it was meant to.

**Comment by JedMeister** [comment]
Opp, just noticed your new post re auFS.

> I did install aufs-dkms dkms linux-headers-amd64 aufs-tools in host and guest.

Did it build the DKMS module and enable it? My guess is that it may have built it, but not enabled it.

Or another possibility is that it did set up auFS correctly, including enabling it. But then when the TKLDev build process used OverlayFS a conflict occurred and the kernel disable auFS.

Assuming that the module is called `aufs`; then you could check to see if it's loaded:
```
lsmod | grep aufs
```
If it's not there, to check that the auFS module exists on your system, try this:
```
find /lib/modules/$(uname -r) -type f -name '*.ko*' | grep aufs
```
If that doesn't return anything, then there's a good chance that the installation didn't build the module. I'm not particularly familiar with `dkms`, but you should be able to list dkms modules with `dkms status`.

If there is definitely an auFS module, then you can try loading it (again assuming it's called `aufs`):
```
modproble aufs
```
If that doesn't error, then you could double check it's loaded by re-running the `lsmod` command noted above.

**Comment by CUCraigT** [comment]
Have now worked through a few more things.

*AUFS*
`dkms status` shows aufs, but all the other commands do not.

*Mapping Directory*
Thanks for the heads up on mounting that. However, I still get an error on the "special files." Doing an `ls -l` on one of those files that gives me the error, you get this:

```
root@tkldev .../products/appliance# ls -l /var/lib/docker/overlay2/067cc3605c5336f8bad46bbb32882dcd389dc0694cdb0599f2bc2def0b089035/diff/var/lib/apt/lists/security.ubuntu.com_ubuntu_dists_xenial-security_universe_source_Sources.lz4
c--------- 1 root root 0, 0 Aug 20 04:00 /var/lib/docker/overlay2/067cc3605c5336f8bad46bbb32882dcd389dc0694cdb0599f2bc2def0b089035/diff/var/lib/apt/lists/security.ubuntu.com_ubuntu_dists_xenial-security_universe_source_Sources.lz4
```

Looks like it is a character device driver with `mknod`, which is the error I get if I try to extract it within the build from a tarball.

Any other thoughts, @JedMeister, or am I going to be stuck using inithooks or some other method to load these at boot time vs. build time? Love the simplicity of this and the ability to build my own appliance; can't find other good options out there. Just can't get it to load those containers, so it works at boot time.

**Comment by CUCraigT** [comment]
Looks like I have it working using `inithooks` to do the initial `docker load`! I am prepopulating the docker images are tar files in an overlay directory. Will detail all of this sometime soon for others that need it.

Also, set up the service and have it enabled.

One quick/easy question for you hopefully... After all the inithooks and everything is run, what is the easy way to either start a service or reboot the system. The service is there as part of overlay in /etc/system.d/system and enabled (via conf.d/main) and working after reboot. Just wanted to start it as well after all the initialization is done.

**Comment by JedMeister** [comment]
Hi @CUCraigT - Sorry I missed your earlier post... TBH, I was out of ideas and didn't really have anything to add really.

As to your question, assuming that you have added a `/usr/lib/inithooks/firstboot.d/20_your_hook` (or similar) then it's just a bash script, so you can do whatever you would from the commandline. E.g.:

```
[...YOUR INITHOOK CONTENTS...]

systemctl restart name-of-your-service
```

Please keep in mind that ideally all inithooks should be idempotent (i.e. won't break things if re-run). If it's not idempotent, then it might be worth writing a file/directory somewhere and check for the existence of that file/directory within your script.

It's also perhaps worth noting that if you don't want your script to be re-run if the user invokes `turnkey-init`, then be sure to put this line near the top of your script:
```
[ -n "$_TURNKEY_INIT" ] && exit 0
```

When you run `turnkey-init` it sets `TURNKEY_INIT=1` - so with the above line, the script won't be re-run. Even if you create/check for a different file, then it's probably still worth considering including that too.

**Comment by JedMeister** [comment]
As I just noted in my edit of the OP, I've just hijacked this issue and repurposed it as a "feature request" for TKLDev docker support... :smile:

**Comment by claflico** [comment]
@JedMeister @CUCraigT Just stumbled across this as I was about to start building a Docker ISO myself. I've been using/managing Docker around 3 years now and there's really not any distros out there anymore for the home user that just wants to get Linux/Docker working without having to configure a bunch of stuff. What's the status on this and what can I do to help?

**Comment by CUCraigT** [comment]
Completely agree with you.

I have this working. I'll reply and send you what I have tomorrow once I'm
back on the computer. I actually have full CI/CD working triggered off
another project creating a release in GitHub. Happy to share and
collaborate.

What are your goals with it?

**Comment by claflico** [comment]
My primary goal is docker, docker-compose, & portainer being pre-installed. After that I have some other docker based projects I'm working on that I want an installable ISO for. Was about to start working on a CentOS 8 custom iso until I read that they were killing off CentOS and knew I needed to look elsewhere.

**Comment by CUCraigT** [comment]
Got it. Happy to share. Didn't publish mine too TKL app store, as it has
our non open source Docker container, but think I have exactly what you
need.

**Comment by CUCraigT** [comment]
@claflico 

At this point, you can install Docker, `docker-compose`, etc as part of the build (shown above). However, what you CANNOT do is actually do a `docker load` or `docker pull` for any images during the build unfortunately, as both Docker and TKL for the build use the `overlay` filesystem. Here was my workaround, and happy to share code; just reach out to me:

1. Use TKL Core as your base image
2. Load Docker, docker-compose
```
echo "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable" >> /etc/apt/sources.list.d/sources.list

curl -fsSL https://download.docker.com/linux/debian/gpg | apt-key add -

apt-get update
apt-get install docker-ce docker-ce-cli containerd.io

echo "Start Docker Service"
systemctl start docker

echo "Install Docker Compose"
curl -L "https://github.com/docker/compose/releases/download/1.26.2/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose
chmod +x /usr/local/bin/docker-compose
```

3. I wanted it to all be self-contained, so as part of a manual step (before build) or automated through CI/CD, I actually get the tarball of the Docker images I want on the box. You can create those on another Linux box (or your build VM) using the `docker save` command. Through a different CI/CD pipeline, I actually do a `docker save` of our image and upload it to Azure Blob storage. Then I pull it down and put it in my overlay directory before doing my build.
4. Create an inithook in `/usr/lib/inithooks/firstboot.d`. I called mine `49atlasity`. In that file, you can do your load at initial boot. You could also do a `docker pull` here instead if you know you have internet access. The two images I am loading are a sqlserver image and our Atlasity image (for reference). `/atlas` is a directory that is in my overlay and has some of my files and where most of the config for my app is, FYI.
```
#!/bin/bash -e
# Extract Docker Images

. /etc/default/inithooks

[ -n "$_TURNKEY_INIT" ] && exit 0

echo Docker Version

docker version > /atlas/version.out

echo Load Docker Images
echo Loading MS SQL Image. This may take awhile.....
docker load -i /atlas/images/sql.tar
echo Loading Atlasity Image. This may take awhile.....
docker load -i /atlas/images/atlas.tar

echo Starting Atlasity Service
systemctl restart atlasity
```

5. This should have you up and going. I also actually set up a service called atlasity to automatically start the docker containers using docker-compose, as well as a python utility to assist with updates, etc.

@JedMeister was a ton of help, and I found some other ways to work around the gaps as of now. However, with this, I can do fully automated builds and then have an ISO file for my customers that has SQL Server and our Atlasity Docker image loaded and VM fully configured and running on their network in ~10 minutes! Which is super cool, thanks to TKL!

Like I said, my GitHub repo is private, but happy to share more or answer any questions. Feel free to reach out to me.

**Comment by JedMeister** [comment]
Nice work @CUCraigT - Thanks for sharing!

@claflico in response to:

> there's really not any distros out there anymore for the home user that just wants to get Linux/Docker working without having to configure a bunch of stuff.

And

> My primary goal is docker, docker-compose, & portainer being pre-installed.

Considering that our whole point is to "get Linux working without having to configure a bunch of stuff" this sounds like your desire to pre-install this software could be the basis of a great official TurnKey appliance candidate. Ultimately, it'd be nice to be able to support the  pre-inclusion of docker images (rather than have to download them on firstboot) but for starters, just having the software pre-installed would be a solid appliance in and of itself...

**Comment by claflico** [comment]
@CUCraigT I had already gotten the first part figured out and working based upon your first post but the second piece is what was needed to finish. I will modify that script to see if it can detect if can connect to the docker hub otherwise it will use the included image. 

Did you have to do anything special for the firewall & docker? I noticed that one of the files has a section for firewall ports.

@JedMeister I can take the first crack at this and at least get it into a repo. Do you want it called just "docker" or "docker-portainer"?

**Comment by CUCraigT** [comment]
@claflico, I tweaked the firewall ports for my application (port 80, 443) to be consistent, but I think it worked fine without doing that. Also, you may have it, but here is code for creating a service using `docker-compose`. Where `WorkingDirectory` is the location of your `docker-compose.yml` file. 
`/overlay/systemd/system/yourservice.service`:
```
# /etc/systemd/system/yourservice.service

[Unit]
Description=Your Service
Requires=docker.service
After=docker.service

[Service]
Type=oneshot
RemainAfterExit=yes
WorkingDirectory=/atlas/system
ExecStart=/usr/local/bin/docker-compose up -d
ExecStop=/usr/local/bin/docker-compose down
TimeoutStartSec=0

[Install]
WantedBy=multi-user.target
```

One other thing you are going to want to do is modify `/overlay/etc/confconsole/confconsole.conf` to specify the `default_nic=eth0`. Otherwise it will mess up the services screen and try to use one of the Docker NICs.

**Comment by claflico** [comment]
Here is my repo that I'm working from for now: https://github.com/claflico/tkl-docker-portainer

@CUCraigT Did you add `/overlay/etc/confconsole/confconsole.conf`? I don't see that in the core repo that I cloned: https://github.com/turnkeylinux-apps/core/tree/master/overlay/etc/confconsole

**Comment by CUCraigT** [comment]
Yep. I just SCPed it from a running one and tweaked it. It is generated if
you don't have it in overlay. Will take a look at your repo shortly.

BTW, the other trick is that you can't put those Docker TAR files in a git
repo as they are too big (beyond hello-world and such).

**Comment by claflico** [comment]
I figured out the confconsole.conf

It's probably taboo in the TKL world but I created a build script that pulls the portainer image into the TKLDev VM, saves and zips it, copies it into overly, and then runs make. I checked that script into the repo instead of the image file.

I tried looking for a Webmin Docker module. If there was one that could be used to manage Docker then I probably wouldn't even bother to include Portainer in the ISO. 

@JedMeister Is there a way to add an "Automatic" option to the partition step? It just automatically sets up the partition instead of having the installer use guided? The use case I have for my next Docker iso is for people who probably aren't going to know or care much about partitioning.

**Comment by CUCraigT** [comment]
Good thoughts. For me, I basically do that same script in my CI/CD
pipeline, so we are thinking on the same page, for sure!
